### ISSUE TITLE
BUG:  Selecting any IGD - TV Control option results in a garbled display

### ISSUE THREAD

**Issue by rtiangha** [Thiblizz/l502x-modded-bios#3]
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

**Description:**

Whenever a TV Control option is selected under IGD Configuration, the display distorts and there's no way to get back to the BIOS options without having to reboot the machine (and any unsaved changes made to the BIOS options up to that point are thus lost).

**Current behaviour:**

A garbled display appears whenever the "VBIOS default" option for the IGD TV Controls is selected.

**Expected behaviour:**

Selecting the items displays what it's supposed to (I assume it's a drop-down list of options), and/or does not produce a garbled display resulting in having to CTRL-ALT-DEL to return to the BIOS options.

**Steps to reproduce the problem:**

1. Go to System Agent (SA) Configuration -> Graphics Configuration -> IGD Configuration 
2. Select "IGD - TV Control" or "IGD TV2 Control" by moving the prompt over the "VBIOS default" option and press the Enter key.
3. The display corrupts and there's no way to get back to the BIOS options without restarting the machine.

**Branch(es):**

Master

**Modded BIOS version:** 1.01-1.03. I never tested with Captain Kasar's modded BIOS version when I used to have that installed. The stock Dell BIOS doesn't have this option displayed.

**GPU model:** GeForce GT 540m on a Dell XPS 15 L502X. I did get the laptop model with the TV tuner card included though, so I don't know if that makes a difference or not.

**Display driver version:** NVIDIA GeForce 378.92 on Windows 10, Nouveau open-source driver in Linux

**Operating system:** Windows 10 Pro 64-bit, Qubes OS (i.e. Fedora 23 based)

**Any related software which may cause the BIOS to act strangely:**

This only happens in the BIOS and the installed operating system has no effect.

[//]: # (This template is for problem reports. For other types of report, edit it accordingly.)
[//]: # (If you are fixing a bug for an existing feature or want to make a change, create a Pull Request.)

**Comment by Thiblizz** [comment]
Hmm, you found a nice one, this is apparently related to the Intel VBIOS.

Had to investigate.

**Comment by rtiangha** [comment]
If that's the case, I wonder if this bug has an effect on some of the other IGD options as well? For example, I run Linux and on Captain Kasar's old modded BIOS, changing the IGD - LCD Panel Type from VBIOS Default to any one of the built-in resolution presets would effect the resolution at which the GRUB boot screen appeared (so I would select an option closer to my display's main resolution, which is 1920x1080, rather than the VBIOS Default).

On these latest versions of the BIOS mod, there is no effect no matter what resolution option is selected under that menu item and thus the GRUB boot screen appears at a really low resolution. It's only cosmetic so it doesn't really affect the functionality of anything, but the low resolution makes it hard to read things on the off chance that I need to change the boot parameters. I could file another bug just for that if you want, or maybe fixing this bug will help that issue too.

**Comment by rtiangha** [comment]
Well, I'm happy to help test. I just don't have access to an external chip programmer so I can't roll back if anything breaks my chip which means I can only test basic things. Let me know if there's anything I can do to help.

Looking forward to and wishing you luck on getting coreboot running on this thing. If it supports Xen virtualization properly so that Qubes OS will run properly on it as well, I'll happily switch to it too. In the meantime, this BIOS mod works mostly fine, except for the IGD quirks I suppose.

**Comment by Thiblizz** [comment]
Yep, thank you for helping, this is always nice to have people like you, still up to test and report problems.

Well for getting coreboot, this is required to desolder the BIOS chip in order to unlock the lock on BIOS regions. I have like ~15 chips and 3 motherboards for the L502X.

I will report back when I will need you to test the IGD bug.

**Comment by rtiangha** [comment]
Thanks. I have one more bug to file; mainly an obscure one that only shows up when running Xen using Linux kernel 4.10 or higher (a mismatch between CPUID and APIC ID) that doesn't really affect anything now, but might in the future if Xen or Linux did stricter checking and enforcement. Some people blame Xen, others blame the hardware (it's only really showing up now because either Xen or the Kernel (can't remember which) is now checking for mismatches), but maybe if it is the hardware advertising the wrong IDs, maybe this BIOS mod can fix things. I'll file that soon just to get it on the radar.

**Comment by Thiblizz** [comment]
Alright, I'm only using Linux, mainly Kali, Arch and Gentoo.

Well my laptop is not working since 2 weeks+ now, waiting for a new motherboard (again), I got ripped off by chinese supplier and so cannot set up anything yet.

But for this one, you will have to open another issues, and the most important, put some logs.

**Comment by rtiangha** [comment]
Actually, upon doing some further research, maybe it'll be fixed in Kernel 4.11 after all:

https://patchwork.kernel.org/patch/9539929/

I just assumed it was a BIOS issue since all the logs said "Firmware Bug." So maybe nothing needs to be done on this end if you're a Xen user and can wait until Kernel 4.11 comes out. I'll hold off on filing any BIOS bugs then until 4.11 comes out and I can test it. Sorry about that!

**Comment by rtiangha** [comment]
Well, I flashed the modded 1.1.0 BIOS with the stock voltages, but there seemed to be no effect. Choosing the VBIOS Default IGD option for TV 1 or TV 2 still results in a garbled display, and changing the resolution option for the IGD - LCD Panel Type does nothing as well. Is the only difference between the two BIOS files the GPU voltage, or does the Modded one have other changes to test?

**Comment by Thiblizz** [comment]
Yes as mentioned in the changelog both **B_MODDED.WPH** and **B_STOCK.WPH** are based on the modded BIOS, only the GPU's core voltage is different (I do not want to have a BIOS that reinclude rootkit).

About the issue referenced here, it seems like the latest VBIOS is problematic and I cannot disassemble it nor have the knowledge to know what is going on.
So I will try to revert back to latest stable VBIOS which does not affect the resolution thing.

**Comment by Thiblizz** [comment]
**v1.1.1** is out, let me know if the problem persists.

**Comment by rtiangha** [comment]
Just flashed v1.1.1. Still has screen corruption with the TV options and changing the resolution under the IGD - LCD Panel Type still has no effect. I didn't try every single option there, though. Only a few, but it still switches off to a low resolution version of GRUB. When GRUB hands off to Linux though, things are set at the proper resolution; that's always happened.

**Comment by Thiblizz** [comment]
I don't think that I will ever fix the TV options problem (as it is not supposed to be available, unlocked != functional).

I'm more focusing on the resolution problem affecting GRUB.

**Comment by Thiblizz** [comment]
I don't think that I will ever fix the TV options problem (as it is not supposed to be available).
(unlocked != functional)

I'm more focusing on the resolution problem affecting GRUB.

By the way, can you test out using **HDMI** and **Mini DisplayPort** to know if they are still working great.

**Comment by Thiblizz** [comment]
Additionally, can you flash these BIOS, "**B_MODDED_TEST1.WPH**" and "**B_MODDED_TEST2.WPH**", I modified the flash tool especially for these two BIOS.

Report me back if any of these fix the problem, I'll really appreciate.

Here is the link of the archive: https://expirebox.com/download/1364f468f9a94be7a5ea35ca10aae0d6.html

**Comment by rtiangha** [comment]
OK, I've downloaded the file and I'm just waiting for Windows 10 to load properly so that I can flash it. Unfortunately, the occurrence of "DXGKRNL_FATAL_ERROR" blue screens/reboots seems to have increased with the last BIOS flash I did. I don't know if this is all related, but one thing I'll try later on (assuming this thing can remain stable enough for me to even flash the BIOS) is switch back to the Dell BIOS and see if I still keep having issues. That way, I'll know if it's my hardware or this BIOS mod. I don't recall having this issues with the final Captain Kasar version. Besides, the Windows 10 Creators Update is coming soon, and so the last thing I want is for the machine to blue screen in the middle of the upgrade. If you can ever find a way to flash a BIOS file outside of Windows again, that'd be great. It'd make things easier for me to test.

I'll let you know the results either way, assuming I can actually flash these things. Am I just checking for the video resolution issue?

**Comment by Thiblizz** [comment]
Wow, I did not made any special changes, and nothing for sure that could result in such problems.

But yeah if you can these two BIOS, that would be great (only to check if the resolution issue get solved).

**Comment by rtiangha** [comment]
Well, I'm hesitant to blame your BIOS for my blue screens; running on "Setup Defaults" seems to have made my Windows 10 a bit more stable for me to test things, so it could have been an option I toggled that is causing it (for example, is it wise to turn stuff like Native PCI Express and Hardware Spread Spectrum Clock on? No idea, but I had them on, among other things. Could have easily been something like that, although I would need more time to test things to narrow it down, which I don't have at the moment.

OK, so I did a battery of tests from the two test versions you provided all the way back to 1.0.1 (the only version I didn't have saved was 1.0.2; you might want to look into using GitHub to tag your releases so they appear on the Releases page; that way people can easily roll back versions if they encounter issues).

The two test versions unfortunately had no effect on the GRUB resolution; changing it from VBIOS Default to something else still results in GRUB outputting a low resolution display until it hands off control to Linux.

The good news is that HDMI and DisplayPort seem to work well across all those BIOS versions (at least under Windows 10; haven't tested under Linux yet), although I didn't test extensively with changing resolutions and such. Basically, all I tested was plugging those into my monitor and see if they outputted at my LCD's native resolution of 1920x1200, which they did.

I'm curious: May I ask what things you're trying to solve the IGD issues? Is it like a mapping thing or something? I notice the Intel VBIOS version in these ones is 2171; I wonder if something more needs to be done to work with that version properly. I believe Kasar's version was 2170, although I didn't reflash his v7 beta to find out in this round of testing.

For now, I've settled on 1.1.1 using the modded voltages and will run Windows 10 for a bit to see how it holds up before switching back to Qubes OS. Thanks again for all your hard work!
